src/Visualization/Force/QuadTree.elm

```diff
@@ -0,0 +1,317 @@
+                                    , nw = singleton toPoint vertex
+                                    , se = qtree
+                                    , sw = Empty
+                                    , ne = Empty
+                                    , aggregate = ()
+                                    }
+
+                            SW ->
+                                Node
+                                    { boundingBox = BoundingBox2d.with { maxX = maxX, maxY = maxY, minX = minX - width, minY = minY - height }
+                                    , sw = singleton toPoint vertex
+                                    , ne = qtree
+                                    , se = Empty
+                                    , nw = Empty
+                                    , aggregate = ()
+                                    }
+
+
+type Quadrant
+    = NE
+    | NW
+    | SE
+    | SW
+
+
+quadrant : BoundingBox2d -> Point2d -> Quadrant
+quadrant boundingBox point =
+    let
+        { minX, minY, maxX, maxY } =
+            BoundingBox2d.extrema boundingBox
+
+        ( midX, midY ) =
+            BoundingBox2d.centroid boundingBox |> Point2d.coordinates
+
+        ( x, y ) =
+            Point2d.coordinates point
+    in
+        if y >= midY then
+            if x >= midX then
+                NE
+            else
+                NW
+        else if x >= midX then
+            SE
+        else
+            SW
+
+
+fromList : (vertex -> Point2d) -> List vertex -> QuadTree () vertex
+fromList toPoint =
+    List.foldl (insertBy toPoint) empty
+
+
+toList : QuadTree x a -> List a
+toList qtree =
+    case qtree of
+        Empty ->
+            []
+
+        Leaf leaf ->
+            let
+                ( first, rest ) =
+                    leaf.children
+            in
+                first :: rest
+
+        Node node ->
+            toList node.nw ++ toList node.ne ++ toList node.se ++ toList node.sw
+
+
+aggregate : Config aggregate vertex -> QuadTree x vertex -> QuadTree aggregate vertex
+aggregate ({ combineAggregates, combineVertices } as config) vanillaQuadTree =
+    case vanillaQuadTree of
+        Empty ->
+            Empty
+
+        Leaf leaf ->
+            let
+                ( first, rest ) =
+                    leaf.children
+            in
+                Leaf
+                    { boundingBox = leaf.boundingBox
+                    , children = ( first, rest )
+                    , aggregate = combineVertices first rest
+                    }
+
+        Node node ->
+            let
+                newNw =
+                    aggregate config node.nw
+
+                newSw =
+                    aggregate config node.sw
+
+                newNe =
+                    aggregate config node.ne
+
+                newSe =
+                    aggregate config node.se
+
+                subresults =
+                    List.filterMap identity
```

Comment:
Nitpick, but this could be `List.filterMap getAggregate [ newNw, newSw, newNe, newSe ]`